### PR TITLE
Add TextDecoder benchmark

### PR DESCRIPTION
**Pull request by ry**
ref #3163 
cc @qwerasd205 

# merge on approval

### CODE REVIEW

**Comment by qwerasd205** [comment]
Since the benchmark utilized random values it should probably be run with a seed to make it deterministic (and thereby more consistent)

**Comment by ry** [comment]
@qwerasd205 I think seeing the distribution is useful

**Comment by qwerasd205** [comment]
Oh, does it do multiple runs?

**Comment by ry** [comment]
@qwerasd205  yes

File: cli/tests/text_decoder_perf.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+function r32() {
+  return Math.floor(Math.random() * 0xffffffff);
+}
+
+function generateRandomUTF8(size) {
```
Comment (piscisaureus) [inline cli/tests/text_decoder_perf.js]: This function is wrong in countless ways.
Care to look again after a good night of sleep?

**Comment by qwerasd205** [comment]
Blame me for the terrible code, as I'm the one who wrote it, it's hastily written while quite tired. I did recognize at the time that it was poor code and suggested replacing the awful function for generating random UTF8 data with a file read.

**Comment by qwerasd205** [comment]
Here's some slightly better code
```ts
tf8 (bytes: number): Uint8Array {
    let randString = '';
    const buffer = new Uint16Array(1024);
    let bufferIndex = 0;
    for (var i = 0; i <= Math.round(bytes / 3); i++) {
        buffer[bufferIndex++] = Math.floor(Math.random() * 0xFFFF) & 0xFC00;
        if (bufferIndex === 1024) {
            randString += String.fromCharCode.apply(null, buffer);
            bufferIndex = 0;
        }
    }
    randString += String.fromCharCode.apply(null, buffer.subarray(0,bufferIndex));
    return new TextEncoder().encode(randString).subarray(0, bytes);
}

const rData = generateRandUtf8(10 * 1024 * 1024);
new TextDecoder().decode(rData);
```
Generates and then decodes 10MB of random well-formed (except the chance of a bad byte or two at the end due to truncation) utf-8 data which consists of entirely 3 or 4 byte sequences. A similar benchmark should probably be made for all ASCII text, and one for just generating the data, to give a reference line for the fixed overhead. Also, somewhat unrelated, but important, there should be unit tests for the TextDecoder and TextEncoder.

**Comment by ry** [comment]
^-- @piscisaureus does that work for you? Note this code is just for benchmarking TextDecode - it's not part of the release.

**Comment by piscisaureus** [comment]
* With the above code, I would bet money that generating the random text takes an order of magnitude longer than decoding it. So this becomes a benchmark of the random text generator, not `TextDecoder`.
* It does not produce well formed UTF-8, since decoding it would yield code points that are not valid unicode, in particular in the range 0xD800-0xDFFF. You can simply test for well-formedness by verifying that `encode(decode(buf))` equals `buf`.
* As a baseline for comparison and general sanity, we should indeed add a benchmark that decodes an all-ASCII buffer, e.g. (`new Uint8Array(size).fill(97)`)

**Comment by qwerasd205** [comment]
>With the above code, I would bet money that generating the random text takes an order of magnitude longer than decoding it. So this becomes a benchmark of the random text generator, not TextDecoder.

Yes, this is why I said
> and one for just generating the data, to give a reference line for the fixed overhead
But, it takes just about as long with the current (very slow) TextDecoder to decode it as it does to generate it, currently.
Though, I still think it would be better just to use a few static test files, rather than trying to generate the data to decode.

>It does not produce well formed UTF-8, since decoding it would yield code points that are not valid unicode, in particular in the range 0xD800-0xDFFF. You can simply test for well-formedness by verifying that encode(decode(buf)) equals buf.

Does TextEncoder ever produce non-well-formed utf-8?

**Comment by piscisaureus** [comment]
> Does TextEncoder ever produce non-well-formed utf-8?

Probably not, so instead you'd end up with a string that contains a lot of replacement characters instead.

**Comment by qwerasd205** [comment]
Yeah I just looked and maybe ~10% of the characters in the text that function outputs are replacement characters. Would you object to benchmarking the decoder on just a string with all sizes of utf-8 character repeated to fill the required size?

**Comment by qwerasd205** [comment]
@piscisaureus 
Would these benchmarks be acceptable?
All 4 byte characters
```ts
const fourBytes = new TextEncoder().encode('😀');

function generateFourBytes (bytes: number): Uint8Array {
    const result = new Uint8Array(bytes);
    for (var i = 0; i < bytes; i++) {
        result[i] = fourBytes[i % 4];
    }
    return result;
}

new TextDecoder().decode(generateFourBytes(10 * 1024 * 1024));
```
Mixed
```ts
const mixed = new TextEncoder().encode('@Ā๐😀');

function generateMixed (bytes: number): Uint8Array {
    const result = new Uint8Array(bytes);
    for (var i = 0; i < bytes; i++) {
        result[i] = mixed[i % 10];
    }
    return result;
}

new TextDecoder().decode(generateMixed(10 * 1024 * 1024));
```
All ASCII
```ts
function generateASCII (bytes: number): Uint8Array {
    return new Uint8Array(bytes).fill(64);
}

new TextDecoder().decode(generateASCII(10 * 1024 * 1024));
```

**Comment by piscisaureus** [comment]
> But, it takes just about as long with the current (very slow) TextDecoder to decode it as it does to generate it, currently.

If this is true then TextDecoder is indeed comically bad. 
But any improvements won't (or would just barely) move the needle because the benchmark overhead would become dominant.

> @piscisaureus
> Would these benchmarks be acceptable?

Yes, much bettter.

Other suggestion: shorter random buffer and decode that 100 times.

**Comment by qwerasd205** [comment]
> Other suggestion: shorter random buffer and decode that 100 times.

Yes, I had that idea, too, because there is some overhead to any decoder, and decoding lots of small stuff is a common use, so representing it would be good. Any suggestion on the length? Maybe 1k?

**Comment by qwerasd205** [comment]
@piscisaureus 
Here's a benchmark that generates 1024 bytes of utf-8 which is a random mix of 4 characters (1 byte long, 2, 3, and 4), then decodes it ten thousand times.

```ts
const mixed = new TextEncoder().encode('@Ā๐😀');

function generateRandom (bytes: number): Uint8Array {
    const result = new Uint8Array(bytes);
    let i = 0;
    while (i < bytes) {
        const toAdd = Math.floor(Math.random()*Math.min(4,bytes-i));
        switch (toAdd) {
            case 0:
                result[i] = mixed[0];
                i++;
                break;
            case 1:
                result[i] = mixed[1];
                result[i+1] += mixed[2];
                i += 2;
                break;
            case 2:
                result[i] = mixed[3];
                result[i+1] = mixed[4];
                result[i+2] = mixed[5];
                i += 3;
                break;
            case 3:
                result[i] = mixed[6];
                result[i+1] = mixed[7];
                result[i+2] = mixed[8];
                result[i+3] = mixed[9];
                i += 4;
                break;
        }
    }
    return result;
}

const randomData = generateRandom(1024);
for (var i = 0; i < 10_000; i++) new TextDecoder().decode(randomData);
```

**Comment by piscisaureus** [comment]
@ry ^-- that works for me

**Comment by qwerasd205** [comment]
It may also be wise to benchmark TextDecoder memory usage, as that's also an issue with the current TextDecoder implementation.

**Comment by ry** [comment]
@piscisaureus updated

@qwerasd205 memory will be checked by the benchmarks. watch for it here after it land https://deno.land/benchmarks.html#max-memory

**Comment by qwerasd205** [comment]
Lint failed, because of `for (var..`. Can just be modified to use `let`. (My use of `for (var..` is a habit I've built as a generic performance consideration, as it used to be the case that `for (let..` loops were significantly slower than `for (var..`. I just checked and it seems like v8 at least has started properly optimizing it so it's equivalent, or sometimes faster. I'll have to start phasing out that habit from my code.)

**Pull request merged**